core/src/main/scala/caliban/execution/Field.scala

```diff
@@ -121,68 +111,70 @@ object Field {
               None
             )
           }
-        case FragmentSpread(name, directives)                           =>
-          val resolvedDirectives = directives.map(directive =>
-            directive.copy(arguments = resolveVariables(directive.arguments, variableDefinitions, variableValues))
-          )
-
-          if (checkDirectives(resolvedDirectives)) {
-            fragments
-              .get(name)
-              .foreach { f =>
-                val t = innerType.possibleTypes
-                  .flatMap(_.find(_.name.contains(f.typeCondition.name)))
-                  .orElse(rootType.types.get(f.typeCondition.name))
-                  .getOrElse(fieldType)
-                if (f.selectionSet.nonEmpty) {
-                  loop(f.selectionSet, t).fields.foreach { field =>
-                    val _field =
-                      if (field._condition.isDefined) field
-                      else
-                        field.copy(
-                          targets = Some(Set(f.typeCondition.name)),
-                          _condition = subtypeNames(f.typeCondition.name, rootType)
-                        )
-                    addField(_field, Some(f.typeCondition.name))
-                  }
+        case fr @ FragmentSpread(name, directives)                      =>
+          val (fields, condition) = memoizedFragments.getOrElseUpdate(
+            fr, {
+              val resolvedDirectives = directives.map(resolveDirectiveVariables(variableValues, variableDefinitions))
+
+              val _fields = if (checkDirectives(resolvedDirectives)) {
+                fragments.get(name).map { f =>
+                  val t = innerType.possibleTypes
+                    .flatMap(_.find(_.name.contains(f.typeCondition.name)))
+                    .orElse(rootType.types.get(f.typeCondition.name))
+                    .getOrElse(fieldType)
```

Comment:
This is the part that concerns me the most, since innerType will be different each time we are here. If we memoize this, it seems that it won't take the new innerType into consideration. But in all cases that I tested this against, innerType.possibleTypes was None, so I'm not sure what needs to be done for it to be Some(...)

Reply:
In test `fragment on union` in `FragmentSpec`, possibleTypes is not None (it contains the subtypes of a union), but `find` doesn't find anything because the fragment is on the root type. Is it even possible? It sounds like even if we removed the find, it would find the type in `rootType.types` anyway, no?

Reply:
That's an interesting point - it seems we should always be able to get the type from `rootType.types`, but I'm not sure if it's possible to get a different type from `innerType.possibleTypes`? Or is that definitely not possible?

Another approach I just thought of if we wanted to be _extremely_ cautious is to move the memoization after L121-122 and only fetch from the cache if those 2 lines return `Some(...)` different value than L123. It does add a bit of overhead but at least this way we can be more certain we're not introducing a regression?

Reply:
I think it only makes sense on inline fragments where you have one fragment for each subtype, but for fragment spreads I think you have to use the parent type anyway, so I think we just go straight to the `orElse`.